packages/bff/src/services/authorizationService.ts

```diff
@@ -0,0 +1,151 @@
+  ORGANIZATION_EXTERNAL_ID_VALUE_CLAIM,
+  ORGANIZATION_ID_CLAIM,
+  SELFCARE_ID_CLAIM,
+  SessionClaims,
+  USER_ROLES,
+  decodeJwtToken,
+  verifyJwtToken,
+} from "pagopa-interop-commons";
+import { genericError } from "pagopa-interop-models";
+import {
+  missingClaim,
+  tenantLoginNotAllowed,
+  tokenVerificationFailed,
+} from "../model/domain/errors.js";
+import { PagoPAInteropBeClients } from "../providers/clientProvider.js";
+import { config } from "../config/config.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function authorizationServiceBuilder(
+  interopTokenGenerator: InteropTokenGenerator,
+  tenantProcessClient: PagoPAInteropBeClients["tenantProcessClient"],
+  allowList: string[]
+) {
+  const readJwt = async (
+    identityToken: string,
+    logger: Logger
+  ): Promise<{
+    roles: string;
+    sessionClaims: SessionClaims;
+    selfcareId: string;
+  }> => {
+    const verified = await verifyJwtToken(identityToken, logger);
+    if (!verified) {
+      throw tokenVerificationFailed();
+    }
+
+    const decoded = decodeJwtToken(identityToken);
+
+    const {
+      success: validToken,
+      data: token,
+      error,
+    } = AuthToken.safeParse(decoded);
+    if (!validToken) {
+      const claim = error?.errors[0]?.path.join(".");
+      throw missingClaim(claim);
+    }
+
+    if (token.role) {
+      throw genericError("User roles in context are not in valid format");
+    }
+
+    const sessionClaims = SessionClaims.parse(token);
+
+    const selfcareId = token[ORGANIZATION].id;
+
+    return {
+      roles: token[USER_ROLES].join(","),
+      sessionClaims,
+      selfcareId,
+    };
+  };
+
+  const assertTenantAllowed = (selfcareId: string, origin: string): void => {
+    if (
+      !config.tenantAllowedOrigins.includes(origin) &&
+      !allowList.includes(selfcareId)
+    ) {
+      throw tenantLoginNotAllowed(selfcareId);
+    }
+  };
+
+  const buildJwtCustomClaims = (
+    roles: string,
+    tenantId: string,
+    selfcareId: string,
+    tenantOrigin: string,
+    tenantExternalId: string
+  ): CustomClaims => ({
+    [USER_ROLES]: roles,
+    [ORGANIZATION_ID_CLAIM]: tenantId,
+    [SELFCARE_ID_CLAIM]: selfcareId,
+    [ORGANIZATION_EXTERNAL_ID_CLAIM]: {
+      [ORGANIZATION_EXTERNAL_ID_ORIGIN_CLAIM]: tenantOrigin,
+      [ORGANIZATION_EXTERNAL_ID_VALUE_CLAIM]: tenantExternalId,
+    },
+  });
+
+  return {
+    getSessionToken: async (
+      correlationId: string,
+      identityToken: string,
+      logger: Logger
+    ): Promise<string> => {
+      logger.info("Received session token exchange request");
+
+      const { sessionClaims, roles, selfcareId } = await readJwt(
+        identityToken,
+        logger
+      );
+
+      const { serialized } =
+        await interopTokenGenerator.generateInternalToken();
+
+      const headers = {
+        "X-Correlation-Id": correlationId,
+        Authorization: `Bearer ${serialized}`,
+      };
```

Comment:
You can now get these from the context, like the logger. If you pass to this function the context from the router, that you can get using `fromBffAppContext`, then in the context you find both the logger and the headers